### ISSUE TITLE
git pre-commit linter should only run on files in the git index

### ISSUE THREAD

**Issue by 6a68** [mozilla/chronicle#322]
The linter is slow because it lints every file in the repo, every time. The point of the git commit hook is to avoid checking in code with syntax errors. So, we should get the list of about-to-be-committed files from git (aka the git index), and only run the linter on those.

**Comment by pdehaan** [comment]
Not sure if we could get something like [**grunt-newer**](https://github.com/tschaub/grunt-newer) to do this for us for the precommit hook, and then have Travis lint everything, just for giggles.

**Comment by 6a68** [comment]
Maybe you can use git-ls-files to get the file list and pass that to the lint task

**Comment by 6a68** [comment]
Ah, nope. The docs in pro git specifically mention using git status --porcelain as a stable version-independent script-friendly listing.

http://unix.stackexchange.com/questions/155046/determine-if-git-working-directory-is-clean-from-a-script

**Comment by pdehaan** [comment]
Quick test w/ **grunt-newer**:
```sh
$ grunt newer:lint
Running "newer:lint" (newer) task
Fatal error: The "newer" prefix is not supported for aliases
```

Possible workaround... create a second "lint" task which explicitly calls `newer:` on each of the subtasks:
```js
  grunt.registerTask('quicklint', [
    'newer:jshint',
    'newer:jscs',
    'newer:jsonlint',
    'newer:copyright'
  ]);
```

Initial testing shows that its about 25% faster (-1 second) than calling `grunt lint`. So I *think* all we'd need to do is change the `lint` npm script from `grunt lint` to `grunt quicklint` and then Travis still runs `grunt lint` for each checkin.

@6a68 Let me know if this sounds do-able, and I can quickly PR it and you can test drive it and see if it is any better for you.
Or if this all sounds lame, we can reconsider if the precommit hook is useful -- or if we should see if we can find a solution that uses `pre-push` hook instead so the linting doesn't happen on every commit.

**Comment by 6a68** [comment]
@pdehaan I was hoping for 99% faster ;-)

How about a counter proposal? There's two steps here, getting the file list from git, and passing it to grunt. I looked into this a little bit:

### getting list of files from git

Here's an example of some things that could happen to a repo:

- Ask git to move a file (auto-adds it into the index): `git mv README.md README.mdz`
- Create a new file and add to the index: `touch foo && git add foo`
- Ask git to delete a file (again, auto-adds the change): `git rm server/bell-oauth-profile.js`
- Edit a file but don't `git add` the changes, let's edit the server/README
- Edit a file and check in the changes, let's edit server/embedly.js
- There's also a dist/ directory that git sees but isn't checked in.

`git status --porcelain` output:
```bash
jareds-partybus:mozilla-chronicle jhirsch$ git status --porcelain
R  README.md -> README.mdz
A  foo
 M server/README.md
D  server/bell-oauth-profile.js
M  server/embedly.js
?? dist/
```
Looks pretty simple to parse: we want to lint the file if the first char on a line is M or A: renamed files don't need to be linted, and neither do deleted files. (I bet there are some less common statuses I've missed, but this should capture most day-to-day dev tasks, and we could read the git-status manpage to be comprehensive.)

There's an npm package called  `simple-git` which attempts to parse this output, but is not quite successful:
```js
Git = require('simple-git')
git = new Git()
git.status(console.log)
// returns:
null { not_added: [ 'dist/' ],
  deleted: [ ',server/bell-oauth-profile.js' ],
  modified: [ 'server/README.md', ',server/embedly.js' ] }
```
Looks like git-status isn't smart enough to handle files being added, and there are some problems with leading commas. I saw a few related packages on npm, but didn't try them; maybe your npm search-fu exceeds mine, and something off-the-shelf can be used. Rolling our own here, though, is like 3 lines of code, plus the bit where we exec the bash command.

### passing list of files to grunt

It seems like we should be able to override the `src` used by the grunt linter tools, and pass in the list of files we want to lint, but I really don't know how. I bet you have ideas, though :-)

**Comment by 6a68** [comment]
oh yeah, there are 9000 hits on 'git' in npm. There has to be a better option than `simple-git`.

**Comment by pdehaan** [comment]
Ppsshh, 99% why not use `--no-verify` flag and get 100% faster!

**Comment by pdehaan** [comment]
I was actually randomly thinking about this today and did find [**husky**](https://github.com/typicode/husky) which seems pretty sweet and lets you easily do `precommit`, `prepush`, and `postmerge` hooks.

It seems to be a lot better at letting you easily do hooks, so maybe moving linting from `precommit` to `premerge` will save you untold minutes per day (if we combine with **grunt-newer** so we're only linting a minimal number of files).

```js
    "precommit": "echo 'hey gurl'",
    "prepush": "npm run lint",
```

Your solution sounds really solid too, I can try Googling and see how easy it'd be to lint an arbitrary list of files. I'd be interesting to see a comparison between your proposed solution and something like **grunt-newer** and see how much time is spent on file globbing/searching vs just Grunt infrastructure.

**Comment by pdehaan** [comment]
@6a68: OK, I submitted a ~~**precommit-hook**~~ --> **husky** PR which will  help by a few milliseconds (see #339).

As for the issue above, I feel better about it, but still need to mentally try and figure out what I'm doing. I found [**staged-git-files**](https://github.com/mcwhittemore/staged-git-files) which seems to be working for me locally and returns a JSON list of [Added, Copied, Deleted, Modified, Renamed, ...] staged files:

```js
'use strict';

var Q = require('q');
var sgf = require('staged-git-files');

// We only care about "Added" or "Modified" files.
var FILTER_STATUSES = ['added', 'modified'];

stagedGitFiles().then(function (results) {
  return filterFiles(results);
}).then(function (files) {
  console.log(JSON.stringify(files, null, 2));
}, function (err) {
  throw err;
});

// Convert the `sgf()` function into a promise. This was blowing up when trying to use
// a clean `Q.denodeify(sgf)`, so we'll do it the ugly verbose way.
function stagedGitFiles() {
  var deferred = Q.defer();
  sgf(function (err, results) {
    if (err) {
      return deferred.reject(err);
    }
    return deferred.resolve(results);
  });
  return deferred.promise;
}

function filterFiles(files) {
  // Filter out unwanted file statuses.
  return files.filter(function (file) {
    var status = file.status.toLowerCase();
    return (FILTER_STATUSES.indexOf(status) !== -1);
  }).map(function (file) {
    // And only return an array of filenames instead of objects.
    return file.filename;
  });
}
```

Basically, it's going to turn this (from **staged-git-files**):
```js
[ { filename: 'grunttasks/stylestats.js', status: 'Added' },
  { filename: 'grunttasks/scsslint.js', status: 'Added' },
  { filename: 'grunttasks/sass.js', status: 'Modified' },
  { filename: 'grunttasks/lint.js', status: 'Modified' },
  { filename: 'grunttasks/htmllint.js', status: 'Added' },
  { filename: 'grunttasks/cssmin.js', status: 'Added' },
  { filename: 'grunttasks/csslint.js', status: 'Added' },
  { filename: 'grunttasks/cmq.js', status: 'Added' } ]
```

... into this:
```json
[
  "grunttasks/stylestats.js",
  "grunttasks/scsslint.js",
  "grunttasks/sass.js",
  "grunttasks/lint.js",
  "grunttasks/htmllint.js",
  "grunttasks/cssmin.js",
  "grunttasks/csslint.js",
  "grunttasks/cmq.js"
]
```

And now I just need to find a way to convert this into an async Grunt task, use **grunt-contrib-copy** to copy that arbitrary list of files into an empty _/.tmp/_ directory, and then update all the lint tasks to have a second target which just lints any JSON or JavaScript in that folder. Then finally use **grunt-contrib-clean** (or whatever) to delete that _/.tmp/_ folder.

THEN we see if that is any faster, or if the ~5 seconds that the `grunt lint` task was taking was mostly consumed by loading the Grunt infrastructure+tasks and not in parsing the files.

**Comment by pdehaan** [comment]
OK, I think this is the general idea which may work, along with some crude `time __` stats (but I can probably get marginally better "per-task" timings using [**time-grunt**](https://github.com/sindresorhus/time-grunt) &mdash; but I'll leave that for another day)...

I called the task "tmplint" since it lints only files in the ".tmp/*" directory, but that sounds lame and isn't very obvious. Maybe we should rebrand it "precommit-lint" or something.

```js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/*
  SGF-Status (git status code):
  - Added (A)
  - Copied (C)
  - Deleted (D)
  - Modified (M)
  - Renamed (R)
  - Type-Change (T) [i.e. regular file, symlink, submodule, etc.]
  - Unmerged (U)
  - Unknown (X)
*/

module.exports = function (grunt) {
  'use strict';

  var path = require('path');

  var Q = require('q');
  var sgf = require('staged-git-files');

  var FILTER_STATUSES = ['added', 'modified'];
  var FOLDER_PREFIX = '.tmp';

  grunt.registerTask('tmplint', function () {
    var done = this.async();
    stagedGitFiles().then(function (results) {
      return filterFiles(results);
    }).then(copyFiles).then(function (files) {
      grunt.task.run('lint:tmp'); // lint the '.tmp/' directory.
      grunt.task.run('clean:tmp'); // delete the '.tmp/' directory.
      done(files);
    }, done);
  });

  /**
   * A promise-based version of the 'staged-git-files' package's main function.
   * @return {Promise} A promise which resolves to an array of staged files.
   */
  function stagedGitFiles() {
    var deferred = Q.defer();
    sgf(function (err, results) {
      err ? deferred.reject(err) :  deferred.resolve(results);
    });
    return deferred.promise;
  }

  /**
   * Convert the array of file objects into an array of filenames.
   * @param  {Array} files [description]
   * @return {Array}       [description]
   */
  function filterFiles(files) {
    return files.filter(function (file) {
      var status = file.status.toLowerCase();
      return (FILTER_STATUSES.indexOf(status) !== -1);
    }).map(function (file) {
      return file.filename;
    });
  }

  /**
   * Copies the staged files in Git to the `FOLDER_PREFIX` directory, and keeps
   * the existing folder/file structure.
   * @param  {Array} files [description]
   * @return {Array}       [description]
   */
  function copyFiles(files) {
    files.forEach(function (file) {
      grunt.file.copy(file, path.join(FOLDER_PREFIX, file));
    });
    grunt.verbose.writeln('copied %d files', files.length);
    return files;
  }
};
```

Also required tweaking a few subtasks to add rules for the ./tmp/ directory. For example:
```js
  grunt.config('jsonlint', {
    app: {
      src: [
        '{.bowerrc,.jshintrc,.jscsrc}',
        '{,app/**/,grunttasks/**/,server/**/,tests/**/}*.json',
        'config/*.{json,example}',
        '!app/bower_components/**'
      ]
    },
    tmp: {
      src: [
        '.tmp/{.bowerrc,.jshintrc,.jscsrc}',
        '.tmp/{,**/}*.json'
      ]
    }
  });
```

Then the "lint" task was tweaked to accept different targets:
```js
  grunt.registerTask('lint', function (target) {
    target = target || 'app';

    grunt.task.run([
      'jshint:' + target,
      'jscs:' + target,
      'jsonlint:' + target,
      'copyright:' + target
    ]);
  });
```

So now you can call `grunt lint:app` or `grunt lint:tmp` and it will call the specified targets for each subtask.

I can tell we're all unimpressed, so here's probably a "best case statistic" for your 99% faster goal (where I have about 8 local Grunt tasks for csslint, htmllint, cmq (combine media queries, cssmin, scsslint, and stylestats):
```sh
$ time grunt lint:app
Running "lint:app" (lint) task

Running "jshint:app" (jshint) task

✔ No problems


Running "jscs:app" (jscs) task
>> 94 files without code style errors.

Running "jsonlint:app" (jsonlint) task
>> 11 files lint free.

Running "copyright:app" (copyright) task

Done, without errors.

real	0m2.680s
user	0m2.481s
sys		0m0.207s
```

And the total time to copy the staged files, run the 4 "lint" subtasks against .tmp/, and then delete the .tmp/ folder:
```sh
$ time grunt tmplint
Running "tmplint" task

Running "lint:tmp" (lint) task

Running "jshint:tmp" (jshint) task

✔ No problems


Running "jscs:tmp" (jscs) task
>> 8 files without code style errors.

Running "jsonlint:tmp" (jsonlint) task
>> 0 files lint free.

Running "copyright:tmp" (copyright) task

Running "clean:tmp" (clean) task
>> 1 path cleaned.

Done, without errors.

real	0m1.308s
user	0m1.102s
sys		0m0.182s
```

So roughly a 50% timing savings, but that may be horribly skewed since timing is hard.

**Comment by pdehaan** [comment]
FWIW, here's the best-_best_ case scenario of setting up the **grunt-newer** module and running it against the `grunt lint:app` targets with zero file changes:
```sh
$ time grunt newerlint
Running "newer:jshint:app" (newer) task
No newer files to process.

Running "newer:jscs:app" (newer) task
No newer files to process.

Running "newer:jsonlint:app" (newer) task
No newer files to process.

Running "newer:copyright:app" (newer) task
No newer files to process.

Done, without errors.

real	0m1.388s
user	0m1.239s
sys 	0m0.150s
```

Or **grunt-newer** with 6 JavaScript file changes:
```sh
...
Done, without errors.

real	0m2.102s
user	0m1.476s
sys 	0m0.262s
```

And my `grunt newerlint` task is basically just:
```js
  // $ npm i grunt-newer -D
  grunt.registerTask('newerlint', [
    'newer:jshint:app',
    'newer:jscs:app',
    'newer:jsonlint:app',
    'newer:copyright:app'
  ]);
```

## CONCLUSION:
My code was a fun experiment, but manually grabbing the staged git files, copying them into a .tmp/ directory, updating any linting tasks to have an "app" and "tmp" target, and then manually `run()`-ing tasks, and deleting files seems like a bit of magic and complexity.
**grunt-newer** will always be faster since it only checks each modified file once instead of each time you re-run lint with those staged files.

**TL;DR:** I think our best bet is the easy integration of **grunt-newer** and using the **husky** module+pre-push hook instead of the **precommit-hook** module+pre-commit hook.

**Comment by pdehaan** [comment]
See #339 for my attempt at replacing **precommit-hook** with **husky** (since that is probably a good move regardless, and moves the linting from pre-commit to pre-push).

**Comment by pdehaan** [comment]
OK, here's my (a) [slightly modified] `grunt lint` task, and (b) the results of my **time-grunt** experiment:

```js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

module.exports = function (grunt) {
  'use strict';

  var SUBTASKS = [
    'jshint',
    'jscs',
    'jsonlint',
    'copyright'
  ];

  grunt.registerTask('lint', function (target) {
    var subtasks = SUBTASKS;

    // If a target was specified, prefix each sub-task name with the specified target value.
    if (target) {
      subtasks = SUBTASKS.map(function (task) {
        return target + ':' + task;
      });
    }

    grunt.task.run(subtasks);
  });
};
```

```sh
$ grunt lint # lint EVERYTHING on EVERY commit (current behavior)

Execution Time (2015-02-17 01:56:06 UTC)
loading tasks   1.2s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 43%
jshint:app     888ms  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 33%
jscs:app       376ms  ▇▇▇▇▇▇▇▇▇▇ 14%
jsonlint:app   226ms  ▇▇▇▇▇▇ 8%
copyright:app   60ms  ▇▇▇ 2%

Total 2.7s
```

```sh
$ grunt lint:newer # only lint newer files with timestamps newer than last cached value.

Execution Time (2015-02-17 01:53:35 UTC)
loading tasks        896ms  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 75%
newer:jshint:app      78ms  ▇▇▇▇ 7%
newer:jscs:app        76ms  ▇▇▇▇ 6%
newer:jsonlint:app    74ms  ▇▇▇▇ 6%
newer:copyright:app   68ms  ▇▇▇ 6%

Total 1.2s
```

Then all we need to do is change the `grunt lint` to `grunt lint:newer` in the pre-commit or pre-push git hooks and we're _magic_! :tada:

**Comment by pdehaan** [comment]
I don't know, maybe it's too weird to try and override the default `target` behavior in the [grunt lint](https://github.com/mozilla/chronicle/issues/322#issuecomment-74606257) task above by turning it into a prefix, and we should probably make some other task which doesn't perform magic. Regardless, the concept and result would be the same.

**Comment by pdehaan** [comment]
Sure, why not... Fixed in #354.
You can use `$ grunt lint` to lint _all the things_, or you can use `$ grunt quicklint` to only lint modified files.
